Sources/SotoCore/Credential/ConfigFileLoader.swift

```diff
@@ -0,0 +1,207 @@
+    struct ProfileConfig: Equatable {
+        let region: Region?
+        let roleArn: String?
+        let roleSessionName: String?
+        let sourceProfile: String?
+        let credentialSource: CredentialSource?
+    }
+
+    /// Profile credential source `credential_source`
+    ///
+    /// Used within Amazon EC2 instances or EC2 containers to specify where the AWS CLI can find credentials to use to assume the role you
+    /// specified with the `role_arn` parameter. You cannot specify both `source_profile` and `credential_source` in the same profile.
+    enum CredentialSource: String, Equatable {
+        case environment = "Environment"
+        case ec2Instance = "Ec2InstanceMetadata"
+        case ecsContainer = "EcsContainer"
+    }
+
+    /// Errors occurring when loading credentials and profile configuration
+    ///
+    /// - missingProfile: If the profile requested was not found
+    /// - missingAccessKeyId: If the access key ID was not found
+    /// - missingSecretAccessKey: If the secret access key was not found
+    enum ConfigFileError: Error, Equatable {
+        case invalidCredentialFileSyntax
+        case missingProfile(String)
+        case missingAccessKeyId
+        case missingSecretAccessKey
+        case missingSourceProfile
+    }
+
+    /// Load shared credentials and profile configuration from passed-in byte-buffers
+    ///
+    /// Credentials file settings have precedence over profile configuration settings
+    /// https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-precedence
+    ///
+    /// - Parameters:
+    ///   - credentialsByteBuffer: contents of AWS shared credentials file (usually `~/.aws/credentials`
+    ///   - configByteBuffer: contents of AWS profile configuration file (usually `~/.aws/config`
+    ///   - profile: named profile to load (usually `default`)
+    ///   - context: credential provider factory context
+    /// - Returns: Credential Provider (StaticCredentials or STSAssumeRole)
+    static func sharedCredentials(from credentialsByteBuffer: ByteBuffer,
+                                  configByteBuffer: ByteBuffer? = nil,
+                                  for profile: String,
+                                  context: CredentialProviderFactory.Context) throws -> CredentialProvider {
+        var config: ConfigFileLoader.ProfileConfig?
+        if let byteBuffer = configByteBuffer {
+            config = try loadProfileConfig(from: byteBuffer, for: profile)
+        }
+        let credentials = try loadCredentials(from: credentialsByteBuffer, for: profile, sourceProfile: config?.sourceProfile)
+
+        // When `role_arn` is defined, temporary credentials must be loaded via STS Assume Role operation
+        if let roleArn = credentials.roleArn ?? config?.roleArn {
+            // If `role_arn` is defined, `source_profile` must be defined too (don't yet support `credential_source`).
```

Comment:
I think the `CredentialProvider` creation should be done by the `CredentialProvider`. `ConfigFileLoader` should just return the details loaded from the file.

Reply:
Agreed, will move.